custom-recipes/object-detection-query-sampler/recipe.py

```diff
@@ -0,0 +1,487 @@
+    scores, top_indices = backend.top_k(scores, k=keras.backend.minimum(max_detections, keras.backend.shape(scores)[0]))
+
+    # filter input using the final set of indices
+    indices             = keras.backend.gather(indices[:, 0], top_indices)
+    boxes               = keras.backend.gather(boxes, indices)
+    labels              = keras.backend.gather(labels, top_indices)
+    other_              = [keras.backend.gather(o, indices) for o in other]
+
+    # Take predictions of all classes
+    scores = keras.backend.gather(cscores, top_indices)
+    
+    # zero pad the outputs
+    pad_size = keras.backend.maximum(0, max_detections - keras.backend.shape(scores)[0])
+    boxes    = backend.pad(boxes, [[0, pad_size], [0, 0]], constant_values=-1)
+    scores   = backend.pad(scores, [[0, pad_size], [0, 0]], constant_values=-1)
+    labels   = backend.pad(labels, [[0, pad_size]], constant_values=-1)
+    labels   = keras.backend.cast(labels, 'int32')
+    other_   = [backend.pad(o, [[0, pad_size]] + [[0, 0] for _ in range(1, len(o.shape))], constant_values=-1) for o in other_]
+
+    # set shapes, since we know what they are
+    boxes.set_shape([max_detections, 4])
+    scores.set_shape([max_detections, int(classification.shape[1])])
+    labels.set_shape([max_detections])
+    for o, s in zip(other_, [list(keras.backend.int_shape(o)) for o in other]):
+        o.set_shape([max_detections] + s[1:])
+
+    return [boxes, scores, labels] + other_
+
+
+class MyFilterDetections(FilterDetections):
+    """ Keras layer for filtering detections using score threshold and NMS.
+    """
+
+    def call(self, inputs, **kwargs):
+        """ Constructs the NMS graph.
+        Args
+            inputs : List of [boxes, classification, other[0], other[1], ...] tensors.
+        """
+        boxes          = inputs[0]
+        classification = inputs[1]
+        other          = inputs[2:]
+
+        # wrap nms with our parameters
+        def _filter_detections(args):
+            boxes          = args[0]
+            classification = args[1]
+            other          = args[2]
+
+            return filter_detections(
+                boxes,
+                classification,
+                other,
+                nms                   = self.nms,
+                class_specific_filter = self.class_specific_filter,
+                score_threshold       = self.score_threshold,
+                max_detections        = self.max_detections,
+                nms_threshold         = self.nms_threshold,
+            )
+
+        # call filter_detections on each batch
+        outputs = backend.map_fn(
+            _filter_detections,
+            elems=[boxes, classification, other],
+            dtype=[keras.backend.floatx(), keras.backend.floatx(), 'int32'] + [o.dtype for o in other],
+            parallel_iterations=self.parallel_iterations
+        )
+
+        return outputs
+
+    def compute_output_shape(self, input_shape):
```

Comment:
I might be missing but this looks exactly like the original [one](https://github.com/fizyr/keras-retinanet/blob/f4017fdfb942ef594bf3f81d846f6e6100fa9e5d/keras_retinanet/layers/filter_detections.py#L184) ?